File: ios/StationToStation/Data/RowGeometry.swift
import CoreGraphics

// MARK: - The drawn row

/// What the Timelines resolution actually draws, as numbers. The iOS half of #116,
/// ported from Android's `RowGeometry.kt` rather than re-derived.
///
/// `rowGeometry` is the one seam. It is arithmetic over the primitives in
/// `Timeline.swift` — `laneStep`, `stripWidth`, `laneXf`, `linesAt`, `nodeHost`,
/// `lineDrawnOffset` — holds no state, touches no Canvas, and needs no device. Three
/// consumers read the same value and none recomputes any part of it: `PeopleRails`
/// strokes it, `LaneGeometryTests` asserts it, and the woven-row dump prints it.
///
/// **The open decision the spec left to this half, settled: the constants are shared.**
/// The tail bend, the stroke weights and the three node radii were listed as
/// Android-only and free to differ here, on the argument that a different point system
/// might want different numbers. They do not differ — this file already drew 56, 2, 1.2
/// and 11/7/5, term for term, because the Canvas was ported from the same source. A
/// SwiftUI point and an Android dp are both 1/163-ish of an inch in practice, so there
/// was never a reason for them to diverge, and two sets of numbers would be two things
/// to keep in step for no gain. If iOS ever wants its own, that is a decision to take
/// deliberately and to write down here.
struct DrawnLine: Equatable {
    /// `Spine` for mine, or a Lane index for a friend's.
    let line: Int
    /// Where the Line enters the row, and where it leaves it into the next one.
    let x: CGFloat
    let toX: CGFloat
    /// Where this row's Node sits down the row.
    let nodeY: CGFloat
    /// The gap the Line leaves at the Node's rim — the Node's outer radius when this
    /// Line was there, and zero when it was not. A Node is a ring you see through and a
    /// Line drawn inside one fills it in; a Line belonging to nobody present is not that
    /// night's business, so it runs straight past without a notch.
    let nodeR: CGFloat
    /// How much of the tail is spent bending toward the next row's x. Never longer than
    /// the room below the Node, or a short row draws its straight stretch backwards
    /// before turning.
    let bendLen: CGFloat
    /// Was this Line at this row.
    let present: Bool
    /// How many Lines lie on this one where it runs through the row, and the weight
    /// that implies.
    let people: Int
    let colour: LineColour
    let width: CGFloat
    /// The same three, for the last stretch of the row — which belongs to the **edge
    /// ahead**, and which *every* Line gets, not only the ones that bend. Company
    /// peeling away leaves a Line alone even when that Line never moves, so a Spine has
    /// to stop reading green there too or it claims a Crossing that ended.
    let peopleAhead: Int
    let colourAhead: LineColour
    let widthAhead: CGFloat
}

/// What a stretch of Line means, not what it looks like. The Canvas resolves a role to
/// a colour; the palette stays where palettes belong, and "more than one Line here means
/// meeting green" becomes assertable without rendering anything.
enum LineColour: Equatable {
    /// More than one Line on this stretch: they *are* one Line and must read as one.
    case meeting
    /// My own Spine. Dimmer on a night I was not at.
    case mine(present: Bool)
    /// A friend's Lane, by Lane index.
    case rail(lane: Int)
    /// A Line running past a night nobody on it attended.
    case absent
}

/// How much of a row's tail is spent bending toward the next Node. A junction belongs to
/// the edge between two Nodes, not to the sliver above the lower one.
let EdgeBend: CGFloat = 56

/// One person walking alone, and what each extra one on the same Line adds.
let LineStrokeWidth: CGFloat = 2
let PerPerson: CGFloat = 1.2

/// The x the Canvas actually strokes my Spine at: half a stroke off `SpineX`, the same
/// nudge the Node marker uses to centre itself on it.
let SpineLineX: CGFloat = SpineX + 1

/// Every Line this row draws, in points.
///
/// `laneWidth` is the strip's current width — the animation runs through here, so a
/// half-open strip is a case with an answer rather than a moment only a screenshot could
/// catch. `rowHeight` is the row's laid-out height; only `bendLen` depends on it.
///
/// A Lane that has not slid into view yet is absent from the result: the strip does not
/// stroke Lines nobody can see.
func rowGeometry(
    _ row: WovenRow,
    _ next: WovenRow?,
    _ lanes: [Friend],
    _ laneWidth: CGFloat,
    _ rowHeight: CGFloat
) -> [DrawnLine] {
    let step = laneStep(lanes.count)
    let strip = stripWidth(lanes.count)
    // Openness is scaled by the Lane count so the Lanes slide out one after another
    // rather than together — which is why a per-Line slide exists at all.
    let open = strip <= 0 ? 0 : min(max(laneWidth / strip, 0), 1) * CGFloat(lanes.count)

    // The geometry asks how many nights a node stands for, never what kind it is: a
    // Section and a Festival are one ring of the same size, drawn once (#166).
    let isFestival = row.node.isSeveral
    let nodeY: CGFloat = isFestival ? 15 : 13
    // The outer radius of whichever Node this row draws. A member gig's smaller ring
    // keeps its proportion by having its own radius, not by scaling the gig's.
    let nodeR: CGFloat = isFestival ? 11 : (row.depth > 0 ? 5 : 7)

    // Every Line is described the same way, mine included. Mine is `Spine` and is only
    // special in that its own Lane is the Spine — so it never slides, and a Crossing I
    // was at happens where I already am.
    let lines = [Spine] + Array(0..<lanes.count)

    func slideOf(_ line: Int) -> CGFloat { line == Spine ? 1 : min(max(open - CGFloat(line), 0), 1) }
    func xOf(_ offset: Int, _ line: Int) -> CGFloat {
        SpineLineX + (laneXf(offset, step) - SpineLineX) * slideOf(line)
    }

    // Was this Line at this row? A membership check on the one which-Line answer, not
    // another open-coded copy of it.
    func thereAt(_ r: WovenRow, _ line: Int) -> Bool { linesAt(r, lanes).contains(line) }

    // How many Lines lie on this one where it runs. Merged Lines are one Line by
    // definition, so without this two of them draw the same path twice and look exactly
    // like one — the weight is what says how many.
    func peopleAt(_ r: WovenRow, _ line: Int) -> Int {
        let here = lineDrawnOffset(r, line, lanes)
        return max(lines.filter { lineDrawnOffset(r, $0, lanes) == here && thereAt(r, $0) }.count, 1)
    }

    // How many travel a bend together: they must share *both* of its ends.
    func peopleAlong(_ to: WovenRow?, _ line: Int) -> Int {
        guard let to else { return peopleAt(row, line) }
        let a = lineDrawnOffset(row, line, lanes)
        let b = lineDrawnOffset(to, line, lanes)
        return max(lines.filter {
            lineDrawnOffset(row, $0, lanes) == a && lineDrawnOffset(to, $0, lanes) == b
        }.count, 1)
    }

    return lines.compactMap { line -> DrawnLine? in
        guard slideOf(line) > 0 else { return nil }

        let x = xOf(lineDrawnOffset(row, line, lanes), line)
        // A row with no next row leaves where it entered: the end of the Spine is
        // defined rather than whatever the loop happened to leave.
        let toX = next == nil ? x : xOf(lineDrawnOffset(next, line, lanes), line)
        let here = thereAt(row, line)
        let gap: CGFloat = here ? nodeR : 0

        let people = peopleAt(row, line)
        let ahead = peopleAlong(next, line)
        return DrawnLine(
            line: line,
            x: x,
            toX: toX,
            nodeY: nodeY,
            nodeR: gap,
            bendLen: max(min((rowHeight - nodeY - gap) * 0.8, EdgeBend), 0),
            present: here,
            people: people,
            colour: roleOf(people, here, line),
            width: widthOf(people),
            peopleAhead: ahead,
            colourAhead: roleOf(ahead, here, line),
            widthAhead: widthOf(ahead)
        )
    }
}

/// Colour follows the geometry rather than the endpoints: where Lines lie on top of each
/// other they *are* one Line and have to read as one, and where one peels away it is
/// alone again and takes its own colour back. Green comes from the *count*, so a Crossing
/// between two friends is green without me being one of them.
private func roleOf(_ people: Int, _ present: Bool, _ line: Int) -> LineColour {
    if people > 1 { return .meeting }
    if line == Spine { return .mine(present: present) }
    return present ? .rail(lane: line) : .absent
}

/// Weight says how many walk this stretch together.
private func widthOf(_ people: Int) -> CGFloat { LineStrokeWidth + PerPerson * CGFloat(people - 1) }

/// Whose Line this is, for a log a person reads.
func lineLabel(_ line: Int, _ lanes: [Friend]) -> String {
    line == Spine ? "spine" : "lane\(line)(\(lanes.indices.contains(line) ? lanes[line].setlistfm : "?"))"
}

File: android/app/src/main/java/io/github/magnusencoded/stationtostation/ui/RowGeometry.kt
package io.github.magnusencoded.stationtostation.ui

import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import io.github.magnusencoded.stationtostation.data.Friend

/**
 * What the Timelines resolution actually draws, as numbers.
 *
 * The whole resolution is geometry — a **Spine** down the page, **Lanes** sliding out
 * beside it, **Lines** bending together at a **Crossing** and peeling away at a
 * **Parting** — and until this existed none of it could be checked except by looking.
 * #69 made *which* **Line** a **Node** sits on one function the canvas calls; this is
 * the same move one step further out, for *where in points*.
 *
 * [rowGeometry] is the one seam. It is arithmetic over the primitives below it
 * ([laneStep], [stripWidth], [laneXf], [linesAt], [nodeHost], [lineOffset]) — it holds
 * no state, touches no canvas, and needs no device. Three consumers read the same
 * value and none recomputes any part of it: the canvas strokes it ([PeopleRails]), the
 * lane geometry suite asserts it, and the woven-row dump prints it ([logWovenRows]).
 *
 * Everything here is in density-independent points. The canvas is the only thing that
 * converts to pixels, which is what makes the same numbers meaningful in a log, in an
 * assertion, and on a platform with a different point system — see `RowGeometry.swift`,
 * which carries these constants term for term.
 */
internal data class DrawnLine(
    /** [Spine] for mine, or a lane index for a friend's. */
    val line: Int,
    /** Where the line enters the row, and where it leaves it into the next one. */
    val x: Dp,
    val toX: Dp,
    /** Where this row's **Node** sits down the row. */
    val nodeY: Dp,
    /**
     * The gap the line leaves at the **Node**'s rim — the node's outer radius when
     * this line was there, and zero when it wasn't. A **Node** is a ring you see
     * through and a line drawn inside one fills it in; a line belonging to nobody
     * present is not that night's business, so it runs straight past without a notch.
     */
    val nodeR: Dp,
    /**
     * How much of the tail is spent bending toward the next row's x. Never longer than
     * the room below the **Node**, or a short row draws its straight stretch backwards
     * before turning.
     */
    val bendLen: Dp,
    /** Was this line at this row. */
    val present: Boolean,
    /** How many **Lines** lie on this one where it runs through the row, and the weight that implies. */
    val people: Int,
    val colour: LineColour,
    val width: Dp,
    /**
     * The same three, for the last stretch of the row — which belongs to the **edge
     * ahead**, and which *every* line gets, not only the ones that bend. Company
     * peeling away leaves a line alone even when that line never moves, so a **Spine**
     * has to stop reading green there too or it claims a **Crossing** that ended.
     */
    val peopleAhead: Int,
    val colourAhead: LineColour,
    val widthAhead: Dp,
)

/**
 * What a stretch of **Line** means, not what it looks like. The canvas resolves a role
 * to a colour; the palette stays where palettes belong, and "more than one **Line**
 * here means meeting green" becomes assertable without graphics in a JVM test.
 */
internal sealed interface LineColour {
    /** More than one **Line** on this stretch: they *are* one line and must read as one. */
    data object Meeting : LineColour

    /** My own **Spine**. Dimmer on a night I wasn't at. */
    data class Mine(val present: Boolean) : LineColour

    /**
     * A friend's **Lane**, by *colour* index — which is their position in the
     * unfiltered lane list, not the lane they are drawn on. The two are the same
     * until someone is hidden; after that the drawn lanes re-pack inward and the
     * colours must not follow them, or hiding one person recolours everyone
     * outside them and a colour you have learned stops meaning a person (#266).
     */
    data class Rail(val colourIndex: Int) : LineColour

    /** A **Line** running past a night nobody on it attended. */
    data object Absent : LineColour
}

/**
 * How much of a row's tail is spent bending toward the next node. A junction belongs
 * to the **edge** between two nodes, not to the sliver above the lower one: with the
 * whole turn crammed into the node's y, a line that steps out for one gig and comes
 * back drew a rounded rectangle instead of parting and rejoining.
 */
internal val EdgeBend = 56.dp

/** One person walking alone, and what each extra one on the same line adds. */
internal val LineWidth = 2.dp
internal val PerPerson = 1.2.dp

/**
 * The x the canvas actually strokes my **Spine** at: half a stroke off [SpineX], the
 * same nudge the node marker in `TimelineItem` uses to centre itself on it.
 */
internal val SpineLineX = SpineX + 1.dp

/**
 * Every **Line** this row draws, in points.
 *
 * [laneWidth] is the strip's current width — the animation runs through here, so a
 * half-open strip is a case with an answer rather than a moment only a screenshot
 * could catch. [rowHeight] is the row's laid-out height; only [DrawnLine.bendLen]
 * depends on it.
 *
 * A **Lane** that has not slid into view yet is absent from the result: the strip does
 * not stroke lines nobody can see.
 *
 * [lanes] is the *visible* lane list ([visibleLanes]), so hiding a person is nothing
 * more than a shorter list — they take no **Lane**, notch no **Node** and are counted
 * into no **Crossing**, with no case written for it anywhere below. [colours] is the
 * other half: the colour index each visible lane keeps, from [laneColours]. Empty
 * means "nobody is hidden", where drawn index and colour index are the same thing.
 */
internal fun rowGeometry(
    row: WovenRow,
    next: WovenRow?,
    lanes: List<Friend>,
    laneWidth: Dp,
    rowHeight: Dp,
    colours: List<Int> = emptyList(),
): List<DrawnLine> {
    val step = laneStep(lanes.size)
    val strip = stripWidth(lanes.size)
    // Openness is scaled by the lane count so the lanes slide out one after another
    // rather than together — which is why a per-line slide exists at all.
    val open = if (strip <= 0.dp) 0f else (laneWidth / strip).coerceIn(0f, 1f) * lanes.size

    // The geometry asks how many nights a node stands for, never what kind it is: a
    // Section and a Festival are one ring of the same size, drawn once (#166).
    val nodeY = if (row.node is TimelineNode.Several) 15.dp else 13.dp
    // The outer radius of whichever node this row draws. A member gig's smaller ring
    // keeps its proportion by having its own radius, not by scaling the gig's.
    val nodeR = when {
        row.node is TimelineNode.Several -> 11.dp
        row.depth > 0 -> 5.dp
        else -> 7.dp
    }

    // Every line is described the same way, mine included. Mine is [Spine] and is only
    // special in that its own lane is the spine — so it never slides, and a crossing I
    // was at happens where I already am.
    val lines = listOf(Spine) + lanes.indices

    // Which colour a drawn lane carries. The one place the two indices are told
    // apart; everything else in this function is geometry and uses the drawn one.
    fun colourOf(line: Int) = colours.getOrElse(line) { line }

    fun slideOf(line: Int) = if (line == Spine) 1f else (open - line).coerceIn(0f, 1f)
    fun xOf(offset: Int, line: Int): Dp = SpineLineX + (laneXf(offset, step) - SpineLineX) * slideOf(line)

    // Was this line at this row? A membership check on the one which-line answer,
    // not another open-coded copy of it.
    fun thereAt(r: WovenRow, line: Int) = linesAt(r, lanes).contains(line)

    // How many lines lie on this one where it runs. Merged lines are one line by
    // definition, so without this two of them draw the same path twice and look
    // exactly like one — the weight is what says how many.
    fun peopleAt(r: WovenRow, line: Int): Int {
        val here = lineOffset(r, line, lanes)
        return lines.count { lineOffset(r, it, lanes) == here && thereAt(r, it) }
            .coerceAtLeast(1)
    }

    // How many travel a bend together: they must share *both* of its ends.
    fun peopleAlong(to: WovenRow?, line: Int): Int {
        if (to == null) return peopleAt(row, line)
        val a = lineOffset(row, line, lanes)
        val b = lineOffset(to, line, lanes)
        return lines.count {
            lineOffset(row, it, lanes) == a && lineOffset(to, it, lanes) == b
        }.coerceAtLeast(1)
    }

    return lines.mapNotNull { line ->
        if (slideOf(line) <= 0f) return@mapNotNull null

        val x = xOf(lineOffset(row, line, lanes), line)
        // A row with no next row leaves where it entered: the end of the spine is
        // defined rather than whatever the loop happened to leave.
        val toX = if (next == null) x else xOf(lineOffset(next, line, lanes), line)
        val here = thereAt(row, line)
        val gap = if (here) nodeR else 0.dp

        val people = peopleAt(row, line)
        val ahead = peopleAlong(next, line)
        DrawnLine(
            line = line,
            x = x,
            toX = toX,
            nodeY = nodeY,
            nodeR = gap,
            bendLen = minOf((rowHeight - nodeY - gap) * 0.8f, EdgeBend).coerceAtLeast(0.dp),
            present = here,
            people = people,
            colour = roleOf(people, here, line, colourOf(line)),
            width = widthOf(people),
            peopleAhead = ahead,
            colourAhead = roleOf(ahead, here, line, colourOf(line)),
            widthAhead = widthOf(ahead),
        )
    }
}

/**
 * Colour follows the geometry rather than the endpoints: where **Lines** lie on top of
 * each other they *are* one **Line** and have to read as one, and where one peels away
 * it is alone again and takes its own colour back. Green comes from the *count*, so a
 * **Crossing** between two friends is green without me being one of them.
 */
private fun roleOf(people: Int, present: Boolean, line: Int, colourIndex: Int): LineColour = when {
    people > 1 -> LineColour.Meeting
    line == Spine -> LineColour.Mine(present)
    present -> LineColour.Rail(colourIndex)
    else -> LineColour.Absent
}

/** Weight says how many walk this stretch together. */
private fun widthOf(people: Int): Dp = LineWidth + PerPerson * (people - 1)

/** Whose line this is, for a log a person reads. */
internal fun lineLabel(line: Int, lanes: List<Friend>): String =
    if (line == Spine) "spine" else "lane$line(${lanes.getOrNull(line)?.setlistfm ?: "?"})"
